floors/rule.go

```diff
@@ -0,0 +1,382 @@
+	SiteDomain string = "siteDomain"
+	PubDomain  string = "pubDomain"
+	Domain     string = "domain"
+	Bundle     string = "bundle"
+	Channel    string = "channel"
+	MediaType  string = "mediaType"
+	Size       string = "size"
+	GptSlot    string = "gptSlot"
+	PbAdSlot   string = "pbAdSlot"
+	Country    string = "country"
+	DeviceType string = "deviceType"
+	Tablet     string = "tablet"
+	Phone      string = "phone"
+)
+
+func getFloorCurrency(floorExt *openrtb_ext.PriceFloorRules) string {
+	floorCur := "USD"
+	if floorExt.Data.Currency != "" {
+		floorCur = floorExt.Data.Currency
+	}
+
+	if floorExt.Data.ModelGroups[0].Currency != "" {
+		floorCur = floorExt.Data.ModelGroups[0].Currency
+	}
+	return floorCur
+}
+
+func getMinFloorValue(floorExt *openrtb_ext.PriceFloorRules, conversions currency.Conversions) (float64, string, error) {
+	var err error
+	var rate float64
+	floorMin := floorExt.FloorMin
+	floorCur := getFloorCurrency(floorExt)
+
+	if floorExt.FloorMin > 0.0 && floorExt.FloorMinCur != "" && floorCur != "" &&
+		floorExt.FloorMinCur != floorCur {
+		rate, err = conversions.GetRate(floorExt.FloorMinCur, floorCur)
+		floorMin = rate * floorExt.FloorMin
+	}
+	return floorMin, floorCur, err
+}
+
+func updateImpExtWithFloorDetails(matchedRule string, imp *openrtb2.Imp, floorVal float64) {
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(`"`+matchedRule+`"`), "prebid", "floors", "floorRule")
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(fmt.Sprintf("%.4f", floorVal)), "prebid", "floors", "floorRuleValue")
+}
+
+func selectFloorModelGroup(modelGroups []openrtb_ext.PriceFloorModelGroup, f func(int) int) []openrtb_ext.PriceFloorModelGroup {
+	totalModelWeight := 0
+
+	for i := 0; i < len(modelGroups); i++ {
+		if modelGroups[i].ModelWeight == 0 {
+			modelGroups[i].ModelWeight = 1
+		}
+		totalModelWeight += modelGroups[i].ModelWeight
+	}
+
+	sort.SliceStable(modelGroups, func(i, j int) bool {
+		return modelGroups[i].ModelWeight < modelGroups[j].ModelWeight
+	})
+
+	winWeight := f(totalModelWeight + 1)
+	debugWeight := winWeight
+	for i, modelGroup := range modelGroups {
+		winWeight -= modelGroup.ModelWeight
+		if winWeight <= 0 {
+			modelGroups[0], modelGroups[i] = modelGroups[i], modelGroups[0]
+			modelGroups[0].DebugWeight = debugWeight
+			return modelGroups[:1]
+		}
+	}
+	return modelGroups[:1]
+}
+
+func shouldSkipFloors(ModelGroupsSkipRate, DataSkipRate, RootSkipRate int, f func(int) int) bool {
+	skipRate := 0
+
+	if ModelGroupsSkipRate > 0 {
+		skipRate = ModelGroupsSkipRate
+	} else if DataSkipRate > 0 {
+		skipRate = DataSkipRate
+	} else {
+		skipRate = RootSkipRate
+	}
+	return skipRate > f(SKIP_RATE_MAX+1)
+}
+
+func findRule(RuleValues map[string]float64, delimiter string, desiredRuleKey []string, numFields int) string {
+
+	ruleKeys := prepareRuleCombinations(desiredRuleKey, numFields, delimiter)
+	for i := 0; i < len(ruleKeys); i++ {
+		if _, ok := RuleValues[ruleKeys[i]]; ok {
+			return ruleKeys[i]
+		}
+	}
+	return ""
+}
```

Comment:
Consider returning a second bool variable for if it was found, similar to how Go does it for map key lookup. This removes the information implicitly carried by the empty string.